src/smt/optimization_solver.h

```diff
@@ -80,38 +85,163 @@ class Objective
+ */
+class OMTOptimizer
+{
+ public:
+  virtual ~OMTOptimizer() = default;
+  /** Given a target node, retrieve an optimizer specific for the node's type
+   * the second field isSigned specifies whether we should use signed comparison
+   * for BitVectors and it's only valid when the type is BitVector
+   **/
+  static std::unique_ptr<OMTOptimizer> getOptimizerForNode(
+      Node targetNode, bool isSigned = false);
+  /** Minimize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> minimize(SmtEngine* parentSMTSolver,
+                                              Node target) = 0;
+  /** Maximize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> maximize(SmtEngine* parentSMTSolver,
+                                              Node target) = 0;
+};
+
+/**
+ * Optimizer for Integer type
+ */
+class OMTOptimizerInteger : public OMTOptimizer
+{
+ public:
+  virtual ~OMTOptimizerInteger() = default;
+  /** Minimize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> minimize(SmtEngine* parentSMTSolver,
+                                              Node target) override;
+  /** Maximize the target node with constraints encoded in parentSMTSolver
+   * Parameters:
+   * - parentSMTSolver: an SMT solver encoding the assertions as the constraints
+   * - target: the target expression to optimize
+   * Return value:
+   * - std::pair<OptResult, Node>: the result of optimization and the optimized
+   *   value
+   **/
+  virtual std::pair<OptResult, Node> maximize(SmtEngine* parentSMTSolver,
+                                              Node target) override;
+
+ private:
+  /** Handles the optimization query specified by objType
+   * (=OBJECTIVE_MINIMIZE/MAXIMIZE) **/
+  std::pair<OptResult, Node> optimize(SmtEngine* parentSMTSolver,
+                                      Node target,
+                                      ObjectiveType objType);
+};
+
+/**
+ * Optimizer for BitVector type
+ */
+class OMTOptimizerBitVector : public OMTOptimizer
+{
+ public:
+  OMTOptimizerBitVector(bool isSigned);
```

Comment:
plz document why this class has a constructor but the others don't (or add one).

Reply:
default constructor is also a constructor.